### PR TITLE
【BAAI】add dlrm training stdcase

### PR DESCRIPTION
**Pull request by yuzhou03**
dlrm nvidia A100  training logs:

[dlrm_nvidia_1x1.zip](https://github.com/FlagOpen/FlagPerf/files/13186577/dlrm_nvidia_1x1.zip)
[dlrm_nvidia_1x8.zip](https://github.com/FlagOpen/FlagPerf/files/13186582/dlrm_nvidia_1x8.zip)
[dlrm_nvidia_1x8_big.zip](https://github.com/FlagOpen/FlagPerf/files/13186586/dlrm_nvidia_1x8_big.zip)
[dlrm_nvidia_2x8_big.zip](https://github.com/FlagOpen/FlagPerf/files/13186590/dlrm_nvidia_2x8_big.zip)

### CODE REVIEW

**Comment by yuzhou03** [comment]
以下目录代码搬运自原始仓库：
FlagPerf/training/benchmarks/dlrm/pytorch/cuda_ext
FlagPerf/training/benchmarks/dlrm/pytorch/dataloaders（除 dataloaders.py以外）
FlagPerf/training/benchmarks/dlrm/pytorch/model/network
FlagPerf/training/benchmarks/dlrm/pytorch/utils

File: training/benchmarks/dlrm/pytorch/config/_base.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
+# DO NOT MODIFY THESE REQUIRED PARAMETERS
+
+# Required parameters
+vendor: str = "nvidia"
```
Comment (shh2000) [inline training/benchmarks/dlrm/pytorch/config/_base.py:4]: none

File: training/benchmarks/dlrm/pytorch/config/_base.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+# Path from which to load a checkpoint
+load_checkpoint_path: str = None
+# Path to which to save the training checkpoints
+save_checkpoint_path: str = None
+
+# =========================================================
+# Saving and logging
+# =========================================================
+# Destination for the log file with various results and statistics
+log_path: str = "./log.json"
```
Comment (shh2000) [inline training/benchmarks/dlrm/pytorch/config/_base.py:98]: 这个是dlrm自己用的吗？可否与perf本身的rank0.out.log合为一个

File: training/benchmarks/dlrm/pytorch/config/_base.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+# Number of initial iterations to exclude from throughput measurements
+benchmark_warmup_steps: int = 0
+
+# =========================================================
+# Machine
+# =========================================================
+# Device to run the majority of the model operations
+base_device = "cuda"
+# If True the script will use Automatic Mixed Precision
+amp = True
```
Comment (shh2000) [inline training/benchmarks/dlrm/pytorch/config/_base.py:113]: 为false时，在不安装amp，不支持torch.cuda.amp的机器上可否运行

File: training/nvidia/dlrm-pytorch/README.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+| 训练吞吐量     | p_train,见“性能指标”    | 不包含每个epoch末尾的评估部分耗时                             |
+| **计算吞吐量** | **p_core,见“性能指标”** | 不包含数据IO部分的耗时(p3>p2>p1)                              |
+| 训练结果       | auc,见“性能指标”        | AUC（Area Under the Receiver Operating Characteristic Curve） |
+| 额外修改项     | 无                      |                                                               |
+
+* 性能指标
+
+| 配置             | precision | fix_hp | e2e_time | p_whole | p_train | p_core | auc    | mem      |
+| ---------------- | --------- | ------ | -------- | ------- | ------- | ------ | ------ | -------- |
+| A100单机8卡(1x8) | amp       | /      | 24413    | 171886  | /       | 186769 | 0.8024 | 8.8/40.0 |
```
Comment (shh2000) [inline training/nvidia/dlrm-pytorch/README.md:47]: image
8卡V100上面，fp32是2.88M，混精是6.9M，请确认下这个0.18M的吞吐量是否计算正确

File: training/benchmarks/dlrm/pytorch/train/trainer.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+        data_stream = torch.cuda.Stream()
+
+        # last one will be dropped in the training loop
+        steps_per_epoch = len(train_dataloader) - 1
+        test_freq = config.test_freq if config.test_freq is not None else steps_per_epoch - 2
+
+        # Accumulating loss on GPU to avoid memcpyD2H every step
+        moving_loss = torch.zeros(1, device=device)
+
+        trainerWrapper = CudaGraphWrapper(model,
```
Comment (shh2000) [inline training/benchmarks/dlrm/pytorch/train/trainer.py:176]: 这个wrapper是用来处理并行的么？还是对FXGraph做了编译，请将这段代码搬到昆仑芯或者天数上尝试一下，确认这个CudaGraphWrapper仅属于pytorch，与nvidia自己的实现或优化无关

File: training/benchmarks/dlrm/pytorch/config/_base.py (hunk 2 above)
Comment (yuzhou03) [reply training/benchmarks/dlrm/pytorch/config/_base.py:98]: 是dlrm训练脚本的Flags里的，perf中没有用到，rm

File: training/nvidia/dlrm-pytorch/README.md (hunk 4 above)
Comment (yuzhou03) [reply training/nvidia/dlrm-pytorch/README.md:47]: dlrm数据集的预处理是4.1. Preprocess to small dataset (FL=15) with Spark GPU，因此是small model. 

单卡，混精是2.09 M. 用的是
`python -m torch.distributed.launch --no_python --use_env --nproc_per_node 8 \
          bash  -c './bind.sh  --cpu=exclusive -- python -m dlrm.scripts.main \
          --dataset /data/dlrm/binary_dataset/ --seed 0 --epochs 1 --amp --cuda_graphs'` 命令启动，用到了bind.sh设置了bind策略。

Comment (shh2000) [reply training/nvidia/dlrm-pytorch/README.md:47]: 2.09单卡和这里的0.18八卡差距仍很大，检查一下计算p_core的方式或单位

Comment (yuzhou03) [reply training/nvidia/dlrm-pytorch/README.md:47]: image
用了deeplearningexample中的dockerfile,构建的镜像，吞吐大幅提升。
image

File: training/benchmarks/dlrm/pytorch/train/trainer.py (hunk 5 above)
Comment (yuzhou03) [reply training/benchmarks/dlrm/pytorch/train/trainer.py:176]: CudaGraphWrapper处理模型并行相关的。
如果不使用cudaGraph, 立即parallelize模型。
如果试用了cudaGraph, 在warmup阶段，在graph stream上，parallelize模型。

Comment (yuzhou03) [reply training/benchmarks/dlrm/pytorch/train/trainer.py]: cuda和apex相关逻辑，已经抽取到nvidia-pytorch/extern目录下了。

File: training/benchmarks/dlrm/pytorch/config/_base.py (hunk 3 above)
Comment (yuzhou03) [reply training/benchmarks/dlrm/pytorch/config/_base.py:111]: kulunxin的机器不支持torch.cuda.amp, 他们提供了torch.cuda.amp的替代torch_xmlir.amp，适配时需要修改代码
天数和mrheads的加速卡是cuda兼容的。

File: training/nvidia/dlrm-pytorch/README.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+| 训练吞吐量     | p_train,见“性能指标”    | 不包含每个epoch末尾的评估部分耗时                             |
+| **计算吞吐量** | **p_core,见“性能指标”** | 不包含数据IO部分的耗时(p3>p2>p1)                              |
+| 训练结果       | auc,见“性能指标”        | AUC（Area Under the Receiver Operating Characteristic Curve） |
+| 额外修改项     | 无                      |                                                               |
+
+* 性能指标
+
+| 配置             | precision | fix_hp | e2e_time | p_whole  | p_train  | p_core   | auc    | mem      |
+| ---------------- | --------- | ------ | -------- | -------- | -------- | -------- | ------ | -------- |
+| A100单机8卡(1x8) | amp       | /      | 2343     | 14384089 | 14835876 | 17601908 | 0.8026 | 10.1/40.0 |
```
Comment (shh2000) [inline training/nvidia/dlrm-pytorch/README.md:47]: 把实验补完

Comment (yuzhou03) [reply training/nvidia/dlrm-pytorch/README.md]: 单机实验数据已补充。

File: training/nvidia/dlrm-pytorch/README.md [hunk 7]
```diff
@@ -45,6 +45,7 @@
 | 配置             | precision | fix_hp     | e2e_time | p_whole  | p_train  | p_core   | auc          | mem       |
 | ---------------- | --------- | ---------- | -------- | -------- | -------- | -------- | ------------ | --------- |
 | A100单机8卡(1x8) | amp       | /          | 2697     | 12503815 | 12908968 | 14514735 | 0.8026/0.802 | 9.4/40.0  |
-| A100单机8卡(1x8) | amp       | bs=2097152 |          | 53983112 | 56040128 | 69826157 | 0.7245/0.802 | 33.4/40.0 |
-| A100单机8卡(1x1) | amp       | bs=262144  |          | 3774285  | 3850633  | 4538679  | 0.5/0.802    | 33.1/40.0 |
+| A100单机8卡(1x8) | amp       | bs=2097152 |          | 53983112 | 56040128 | 69826157 |              | 33.4/40.0 |
```
Comment (shh2000) [inline training/nvidia/dlrm-pytorch/README.md:48]: 八卡比16卡快么

File: training/benchmarks/dlrm/pytorch/config/_base.py (hunk 3 above)
Comment (shh2000) [reply training/benchmarks/dlrm/pytorch/config/_base.py:111]: amp为false时是否可运行？

Comment (yuzhou03) [reply training/benchmarks/dlrm/pytorch/config/_base.py:111]: A100上amp和cuda_graphs同时设置为false，可以运行。只amp=False，会报错。
对于kulunxin的机器不支持torch.cuda.amp，适配时需要在extern覆盖stdcase的实现。

File: training/nvidia/dlrm-pytorch/README.md (hunk 7 above)
Comment (yuzhou03) [reply training/nvidia/dlrm-pytorch/README.md:48]: 理论上来说，多机训练比单机训练能提高吞吐量。

与单机训练相比，多机间的通信开销以及数据分布是否均衡会影响吞吐量。

1。本case的数据集为24天的点击日志，预处理后的数据集有363G，转成JointSparseEmbedding的tensor较大。

2。在多机环境中，数据的分布对于训练的吞吐量也很重要。如果数据在不同机器之间的分布不均匀，可能导致一些机器等待其他机器的数据，从而减缓整个训练的速度。
这是原始仓库的训练脚本运行时的各卡的显存占用情况，各卡之间的数据分布并不是很均匀。
image

Comment (shh2000) [reply training/nvidia/dlrm-pytorch/README.md:48]: 而且单机比单卡快了15倍。你用SDP了吗？为什么加速比可以超过8.0

File: training/benchmarks/dlrm/pytorch/config/_base.py (hunk 3 above)
Comment (shh2000) [reply training/benchmarks/dlrm/pytorch/config/_base.py:111]: 同时false的吞吐量正常吗，是否为amp的50%-70%，通常32bit是amp的0.5-0.7倍吞吐量

File: training/nvidia/dlrm-pytorch/README.md (hunk 7 above)
Comment (yuzhou03) [reply training/nvidia/dlrm-pytorch/README.md:48]: 没有用SDP. 1x8和1x1 bs相差8倍。

**Review by shh2000 — APPROVED** [review]

**Review by upvenly — APPROVED** [review]
